[clx3210] iP

clx3210 opened the pull request:
### DukePro (Olivero)
> “Your mind is for having ideas, not holding them.” – David Allen

This bot frees your mind of clutter by managing your tasks. It's:

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/clx3210/ip/releases/tag/A-Jar).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :wink:

And it's **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can practice it too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Contribute by cloning the repo (`git clone https://github.com/clx3210/ip.git`).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

clx3210 committed:
Rename main class

clx3210 committed:
Add chatbot greeting and exit

clx3210 committed:
Modularise chatbot messages

clx3210 committed:
Add echo functionality

clx3210 committed:
Add ability to add and list text task inputs

clx3210 committed:
Refactor text inputs into task objects

clx3210 committed:
Add ability to mark and unmark task done status

clx3210 committed:
Add classes for todo, event and deadline tasks

clx3210 committed:
Reformat files to use tab space 4

clx3210 committed:
Add support for tracking ToDos, Deadline and Event tasks

clx3210 committed:
Add automated testing

clx3210 committed:
Add custom error classes

clx3210 committed:
Add edge case checks to task object methods

clx3210 committed:
Implement basic error handling for incorrect user inputs

clx3210 committed:
Add support for deleting tasks

clx3210 committed:
Update expected test case

clx3210 committed:
Refactor switch cases to use command type enums

clx3210 committed:
Update Readme.md instructions

clx3210 committed:
Rename 'Task' parsing to 'Command' parsing

clx3210 committed:
Add parser for tasks

clx3210 committed:
Make tasks formattable

clx3210 committed:
Add ability to save tasks on disk

clx3210 committed:
Merge branch 'branch-Level-7'

clx3210 committed:
Add date parser utility class

clx3210 committed:
Improve date representation

clx3210 committed:
Add stricter datetime checks

clx3210 committed:
Merge branch 'branch-Level-8'

clx3210 committed:
Add 'Ui', 'Storage' and 'Responses' classes

clx3210 committed:
Modularize bot startup code

clx3210 committed:
Modularize bot responses

clx3210 committed:
Add command classes and command parser

clx3210 committed:
Refactor classes to use composition

clx3210 committed:
Move tasklist load and save logic into storage class

clx3210 committed:
Improve error message abstraction

clx3210 committed:
Fix invalid task number bug

clx3210 committed:
Add todo message usage

clx3210 committed:
Merge branch 'branch-A-MoreOOP'

clx3210 committed:
Move source files into related packages

clx3210 committed:
Update test cases

clx3210 committed:
Update README.md instructions

clx3210 committed:
Merge branch 'branch-A-Packages'

clx3210 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

clx3210 committed:
Update Gradle config

clx3210 committed:
Add unit test cases

clx3210 committed:
Fix string response formatting

clx3210 committed:
Update Gradle config

clx3210 committed:
Merge branch 'branch-A-JUnit'

clx3210 committed:
Add bot response for missing save file

clx3210 committed:
Add Javadoc for public members

clx3210 committed:
Add check style configs

clx3210 committed:
Fix non-Javadoc style errors

clx3210 committed:
Add find command

clx3210 committed:
Merge branch 'branch-A-JavaDoc'

clx3210 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	build.gradle
#	src/main/java/olivero/commands/CommandType.java
#	src/main/java/olivero/commands/DeadlineCommand.java
#	src/main/java/olivero/commands/EventCommand.java
#	src/main/java/olivero/commands/MarkCommand.java
#	src/main/java/olivero/common/Responses.java
#	src/main/java/olivero/parsers/DateParser.java
#	src/main/java/olivero/parsers/Parser.java
#	src/main/java/olivero/parsers/TaskParser.java
#	src/main/java/olivero/storage/Storage.java
#	src/main/java/olivero/tasks/Deadline.java
#	src/main/java/olivero/tasks/Event.java
#	src/main/java/olivero/tasks/ToDo.java
#	src/test/java/olivero/tasks/TaskListTest.java

Conflict resolution is to mainly include both changes since changes
on master are mainly Javadoc additions while the incoming branch
contains code quality improvements.

clx3210 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/olivero/commands/CommandType.java
#	src/main/java/olivero/parsers/Parser.java
#	src/main/java/olivero/tasks/TaskList.java
#	src/main/java/olivero/ui/Ui.java

Conflict resolution: include both

clx3210 committed:
Fix code style issues

clx3210 committed:
Integrate javafx into existing user interface

clx3210 committed:
Merge branch 'branch-Level-10'

clx3210 committed:
Fix checkstyle issues

clx3210 committed:
Merge branch 'branch-A-Checkstyle'

clx3210 committed:
Fix incorrect error throwing

iuhiah commented on src/main/java/olivero/Olivero.java:
```diff
@@ -0,0 +1,70 @@
+import olivero.storage.Storage;
+import olivero.tasks.TaskList;
+
+/**
+ * Represents the main entry point for the program.
+ */
```
All the JavaDocs are descriptive and easy to understand :)

clx3210 replied:
Thanks!

iuhiah commented on src/main/java/olivero/common/Responses.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Common display responses that will be shown to the user.
+ */
+public final class Responses {
```
Really like this idea of having a fixed set of responses, might implement this in my own version :>

iuhiah commented on src/main/java/olivero/tasks/Task.java:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @return Task completion status.
+     */
+    public boolean isDone() {
```
```suggestion
    public boolean getIsDone() {
```

Don't think this follows getter naming conventions, should be getIsDone()!

clx3210 replied:
Hi, since this is a boolean method, it has been named to sound like a boolean based on the coding conventions

iuhiah commented on src/main/java/olivero/tasks/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+    public String toFormattedString() {
+        int doneStatus = isDone() ? 1 : 0;
+        String dateString = DateParser.asInputDateString(endDate);
+        return "D" + " | " + doneStatus + " | " + getDescription() + " | " + dateString;
```
```suggestion
        return "D" + " | " + super() + " | " + dateString;
```

I would recommend extracting `doneStatus + " | " + getDescription()` in the subclasses to the Task superclass and calling super() instead to maintain the abstraction barrier (more so for the isDone attribute since it does not seem to be accessed anywhere else)

iuhiah commented on src/test/java/olivero/parsers/ParserTest.java:
```diff
@@ -0,0 +1,232 @@
+
+    @Test
+    public void parse_blankArguments_exceptionThrown() {
+        String expected = "W-WHAT?! "
+                + "I do not understand what you just said :(";
```
```suggestion
        String expected = Responses.RESPONSE_UNKNOWN_COMMAND
```

Is there any reason the expected responses are not imported from common.Responses? It would help if you had to make changes/add to the responses along the way (less lines to edit)

clx3210 replied:
This helps to keep the testing code independent of the source code, to prevent false negatives

iuhiah commented on src/main/java/olivero/parsers/TaskParser.java:
```diff
@@ -0,0 +1,92 @@
+        }
+        String[] lines = rawTasks.split(System.lineSeparator());
+        for (String line : lines) {
+            String[] args = line.split(Pattern.quote(SEPARATOR));
```
```suggestion
            String[] args = line.split(Pattern.quote(SEPARATOR), limit);
```

Consider taking into account cases where the task description has '|' with split()'s limit argument

iuhiah commented on src/main/java/olivero/parsers/Parser.java:
```diff
@@ -0,0 +1,196 @@
+/**
+ * Parses commands from the user input and returns a {@code Command} object.
+ */
+public class Parser {
```
```suggestion
public class InputParser {
```

Maybe renaming this to InputParser (or something similar) would better follow the naming of the other parsers, I assumed that this was the superclass of DateParser and TaskParser

iuhiah reviewed (commented):
Code is relatively easy to read and understand, do remember to stick to the style conventions (remove all the unnecessary linebreaks)! Class abstraction feels slightly overwhelming, I would think about extracting common methods to existing classes (e.g. the methods in CommandResult could be absorbed into the main Command class and have the subclasses override the method as needed).

Nit: TaskList tests are too long IMO, particularly the methods that test the adding of a lot of tasks. Testing adding one of each type of task would be sufficient if the test goal was to add multiple tasks as a single string input.

clx3210 committed:
Modularize command parsers

The original parser contains too many responsibilities for the
various commands. Thus we split each parsing responsibility
into a given sub class related to the particular command
being parsed.

clx3210 committed:
Add Javadoc to parsers

niha81 commented on src/main/java/olivero/parsers/DateParser.java:
```diff
@@ -0,0 +1,65 @@
+     * @return The formatted string of the given date.
+     * @throws DateTimeException If an error occurs during formatting
+     */
+    public static String asDisplayDateString(LocalDateTime date) {
```
Perhaps the method could be named something like 'formatForDisplay'?

niha81 commented on src/main/java/olivero/parsers/DateParser.java:
```diff
@@ -0,0 +1,65 @@
+     * @return The formatted string of the given date.
+     * @throws DateTimeException If an error occurs during formatting
+     */
+    public static String asInputDateString(LocalDateTime date) {
```
Similar to the method above maybe the name could be changed to 'formatForInput'

niha81 reviewed (commented):
LGTM! I like that your method and variable names are very clear it really helps with understanding the code : )

clx3210 committed:
Refactor code quality

This commit removes redundant imports from the existing classes,
updates naming and code formatting in order to better comply
with the Java coding standards.

clx3210 committed:
Merge branch 'branch-improve-parsing'

clx3210 committed:
Add code assertions

The original code base lacks assertion checks to maintain key
program invariants, making the programmer prone to hidden mistakes.

This commit adds assertion checks particularly in command
parsing sections to ensure correct program behavior.

clx3210 committed:
Refactor code quality

clx3210 committed:
Modularize task parsers

This commit splits task parsing into subclasses based on
the type of task being decoded, to improve readability and
SLAP.

Additionally storage now composes encoder and decoder classes
to handle task formatting while allowing storage to focus on
file IO responsibility.

clx3210 committed:
Extract some validations into utility class

Adds a utility class for validation while parsing commands
to reduce code duplication.

clx3210 committed:
Refactor code base

Extract task parsing routines into utility class to improve
readability and code simplicity. Removed unused code and
moved most magic literals into constants to improve code quality.

clx3210 committed:
Indicate stream implementation

clx3210 committed:
Merge pull request #2 from clx3210/branch-A-CodeQuality

Improve code quality

clx3210 committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/olivero/commands/CommandResult.java
#	src/main/java/olivero/commands/DeleteCommand.java
#	src/main/java/olivero/commands/FindCommand.java
#	src/main/java/olivero/commands/MarkCommand.java
#	src/main/java/olivero/commands/UnMarkCommand.java
#	src/main/java/olivero/parsers/commands/MarkCommandParser.java
#	src/main/java/olivero/parsers/commands/UnMarkCommandParser.java

Conflict resolution: include both and fix syntax errors

clx3210 committed:
Merge branch 'master' into branch-A-Streams

clx3210 committed:
Merge pull request #1 from clx3210/branch-A-Assertions

Add code assertions

clx3210 committed:
Merge branch 'master' into branch-A-Streams

clx3210 committed:
Merge pull request #3 from clx3210/branch-A-Streams

Add streams

clx3210 committed:
Add Gradle workflow file

clx3210 committed:
Implement mass-ops for mark, unmark and delete commands

clx3210 committed:
Fix check style issues

clx3210 committed:
Merge branch 'branch-C-MassOps'

clx3210 committed:
Remove text-ui-test files

clx3210 committed:
Fix naming issues

clx3210 committed:
Merge pull request #4 from clx3210/branch-cleanup

Housekeep code

clx3210 committed:
Remove window close delay for bye command

clx3210 committed:
Improve code quality

clx3210 committed:
Remove redundant description parsing logic

clx3210 committed:
Add user guide

clx3210 committed:
Add more unit tests

clx3210 committed:
Add reuse comments

clx3210 committed:
Upgrade Gui

clx3210 committed:
Fix task indexing bug

clx3210 committed:
Update user guide

clx3210 committed:
Update product image

clx3210 committed:
Update README.md

clx3210 committed:
Improve naming and unit test code quality

clx3210 committed:
Rename todo

clx3210 committed:
Update user guide

clx3210 committed:
Merge pull request #6 from clx3210/branch-cleanup-code

More clean up

clx3210 committed:
Update README.md

clx3210 committed:
Improve code quality

Add reserve token check for task descriptions.

Improve regex for various commands.

clx3210 committed:
Update readme

clx3210 committed:
Tweak user guide

clx3210 committed:
Fix docs rendering

clx3210 committed:
Update README.md again

clx3210 committed:
Update README.md
